### PR TITLE
Toptica laser locking and scanning experiments

### PR DESCRIPTION
**Pull request by mkbhaskar**
Implemented functionality for using Toptica laser, wavemeter, and Swabian instruments counter for initial laser scanning experiments

## Toptica laser control

Implemented a basic driver and client-server functionality for the toptica laser, using the `telnetlib` communication library which can connect to the Toptica DLC-Pro via ethernet. See `pylabnet.hardware.lasers.toptica` for details. In addition to laser locking capability (see below), easy laser on/off functionality from launch control has been added (see `pylabnet.launchers.toptica_on` and `pylabnet.launchers.toptica_off`)

## Laser locking 

Tested fully networked laser locking. Highlights
- Can run the wavemeter software and server on the computer in B16, and all other scripts accessing wavemeter data from other computers in the network. 
- Added a `toptica_lock` launcher with dedicated config file for opening the wavemeter software to lock the toptica laser (PID parameters pre-tuned)
- Added convenience function for centering the piezo voltage (`Zero Voltage` button on GUI), helpful when the laser hits a mode-hop point

Example of laser locked and stable (good PID)

image

Example of laser scanning (manually using the GUI)

image (1)

## Counting

Modified counter drivers for Swabian Instruments for practical experiment needs:
- Upgraded the `pylabnet.hardware.counters.swabian_instruments.cnt_monitor` driver to support simultaneous acquisition of multiple measurements (e.g. continuous counter monitoring and laser scan + counting)
- Implemented a new counting functionality based on the `Countrate` class in the `TimeTagger` API, giving reliable time-normalized measurements of count rates useful for laser scanning

## Laser scan and count experiment

Implemented and tested the laser scan and counting experiment based on the `pylabnet.utils.sweeper` module. This allows for very simple encoding of the experiment:
```python3
def experiment(frequency):

    # Set frequency
    wm.update_parameters([{
        'channel': 6, 'setpoint': frequency
    }])

    return ctr.get_count_rate(name='sweeper', integration=0.1)
```
The frequency parameter can then be swept, and the data will be plotted, and optionally saved. This is what data ends up looking like

image

This can all happen in parallel with the normal wavemeter GUI application as well as continuous counter monitoring applications, *all with the same hardware, not by using separate physical channels*. This point is a nice highlight of our modular software platform, as compared to other implementations, which would require dedicated hardware channels for each application (wavemeter monitoring, laser locking, laser scanning, continuous count monitoring, and experiment/scanning count acquisition).

### CODE REVIEW

**Comment by thtrummer** [comment]
There is an official [Python SDK](https://pypi.org/project/toptica-lasersdk/) for the DLC pro (if you don't mind external dependencies).

**Comment by cknaut** [comment]
Tested laser-on, laser-off, and locking, works great. Maybe we should implement the laser-on and laser-off functionality as a `StaticLine` class? This would probably require adding the `is_staticline_high/low` check, which I wanted to implement anyhow and will enable us to use devices which use TTL switching, but also have a status pin (like the non-Thorlabs shutters). 

Tested the scanning functionality in the `laser_scan_closed_loop` notebook. Scanning works as expected:
image

But, for some reason, no scan output is produced:
image

Let's have a look at this later, very likely I'm using the code wrong.

**Comment by mkbhaskar** [comment]
> Let's have a look at this later, very likely I'm using the code wrong.

For reference, it's an issue with rendering widgets in jupyter notebook. probably need to change some notebook settings. It works well embedded in VSCode

File: pylabnet/hardware/staticline/staticline.py [hunk 1]
```diff
@@ -150,6 +150,12 @@ def _setup_NiDaqMxDriver(self, **kwargs):
         # Log successfull setup.
         self.log.info(f"NiDaq {self.hardware_module.dev} output {ao_output} successfully assigned to staticline {self.name}.")
 
+    def _setup_toptica(self, **kwargs):
+
+        self.up = lambda: self.hardware_module.turn_on()
+        self.down = lambda: self.hardware_module.turn_off()
```
Comment (mkbhaskar) [inline pylabnet/hardware/staticline/staticline.py:156]: @cknaut will this work? should we try writing a launcher for all relevant static-lines for the current experiment (what else is there, just the imaging shutter?), and testing?

Comment (cknaut) [reply pylabnet/hardware/staticline/staticline.py:156]: Almost, just need to register the driver module name in the following line in the `init` function.

```python

        # Dictionary listing all hardware modules which can address
        # staticlines and their corresponding setup functions.
        registered_staticline_modules = {
            'zi_hdawg':  self._setup_HDWAGDriver,
            'nidaqmx_card': self._setup_NiDaqMxDriver
        }
```

Problem with this staticline class is that it cannot check if the staticline-toggle was successful, which in turn can yield a scenario where the position of the button is not indicative of the actual state of the staticline. Added an issue (#82).

**Comment by mkbhaskar** [comment]
> Almost, just need to register the driver module name in the following line in the init function.

Ok, done in 3b946a1 - still needs to be tested

> Problem with this staticline class is that it cannot check if the staticline-toggle was successful, which in turn can yield a scenario where the position of the button is not indicative of the actual state of the staticline. Added an issue (#82).

We definitely want this functionality, but it shouldn't lead to any errors in turning on/off static lines right? It just means we will not be able to use the static line class to check the current status of a line.

**Comment by mkbhaskar** [comment]
Will leave the static line issue for #82 and close/merge this PR

**Pull request merged**